ssm_cache/cache.py

```diff
@@ -5,15 +5,77 @@
 from functools import wraps
 import six
 
-import boto3
-
 class InvalidParameterError(Exception):
     """ Raised when something's wrong with the provided param name """
 
 class Refreshable(object):
-    """ Abstract class for refreshable objects (with max-age) """
+    """ Abstract class for refreshable objects (with max-age)
+
+    The class provides methods get_boto3_client() and get_boto3_resource() that cache
+    the clients/resources in the class, reducing overhead in the Lambda invocations.
+    These also rely on the get_boto3_session() method, which in turn uses
+    BOTO3_SESSION_FACTORY if it is set, allowing overriding with mock sessions for
+    testing. Similarly, BOTO3_CLIENT_FACTORY and BOTO3_RESOURCE_FACTORY, both of which
+    can be set to callables that take a session and a name, can be set to override
+    client and resource creation.
+
+    Some hooks are provided to override behavior. These are class fields, since they are called
+    by a class method.
+    * BOTO3_SESSION_FACTORY takes no input and returns an object that acts like a boto3 session.
+        If this class field is not None, it is used by get_boto3_session() instead of creating
+        a regular boto3 session. This could be made to use placebo for testing
+        https://github.com/garnaat/placebo
+    * BOTO3_CLIENT_FACTORY takes the AWS service name (as defined in boto3) as input and returns
+        an object that acts like a boto3 client. If this class field is not None, it is used by
+        get_boto3_client() instead of creating a regular boto3 client.
+    """
+
+    @classmethod
+    def default_session_factory(cls):
+        """Default session factory that creates a boto3 session."""
+        import boto3
+        return boto3.session.Session()
+
+    @classmethod
+    def default_client_factory(cls, session, name):
+        """Default client factory that creates a client from the provided session."""
+        return session.client(name)
+
+    BOTO3_SESSION_FACTORY = default_session_factory
+    BOTO3_CLIENT_FACTORY = default_client_factory
+
+    BOTO3_SESSION = None
+    BOTO3_CLIENTS = {}
+
+    @classmethod
+    def get_boto3_session(cls):
+        """Use the defined session factory to create an object that acts like a boto3 session.
+        Defaults to boto3.session.Session(); set BOTO3_SESSION_FACTORY to inject a different
+        session factory.
+
+        You should not need to call this method yourself; it is meant for internal use."""
```

Comment:
If the method isn't meant to be public, I would rename it into `_get_boto3_session` and remove this comment (as for Python conventions)

Reply:
👍